### PR TITLE
seccomp: conditionally add socketcall() based on system and base

### PR DESCRIPTION
**Pull request by jdstrand**
The socketcall() syscall is an older multiplexing syscall that is historically
used on some architectures (eg, i386) for various socket related operations
like socket(), bind(), listen(), etc. socketcall()'s 'args' cannot be mediated
with seccomp, so not allowing the syscall where possible is desirable. The
current implementation is fine on architectures that don't support socketcall()
at all (eg, amd64, ARM), but socketcall() should only be add conditionally on
the others since it can be used instead of the mediated syscalls. LP: #1446748.

For i386, the 4.3 kernel added individual syscalls for all the multiplexed
socketcall() operations and glibc was modified to prefer the individual
syscalls over socketcall() (statically compiled programs need to have been
built with the newer glibc). With a new enough kernel and glibc, on i386 we
can remove socketcall() since glibc will try the regular syscalls first, then
fallback to socketcall(), thus allowing proper mediation for this family of
syscalls.

As such, snapd will determine if the system requires the use of socketcall().
Factors:
- if the kernel architecture is amd64, armhf or arm64, do not require
  socketcall (unused on these architectures)
- if the kernel architecture is i386 or s390x:
  - if the kernel is < 4.3, force the use of socketcall()
  - for backwards compatibility, if the system is Ubuntu 14.04 or lower,
    force use of socketcall()
  - for backwards compatibility, if the base snap is unspecified, "core" or
    "core16", then force use of socketcall()
  - otherwise (ie, if new enough kernel, not 14.04, and a non-16 base
    snap), don't force use of socketcall()
- if the kernel architecture is not any of the above, force the use of
  socketcall()

IMPORTANT: currently this PR hardcodes the base snaps to consider
forcing use of socketcall(). This means that the list of base snaps needs
to be hard-coded in snapd. Since new base snaps (eg, Fedora, Debian,
Ubuntu derivatives (even 16.04-based ones), etc) will tend to have a
newer glibc and because base snaps are tightly controlled in the store,
this seems like the best default for the initial implementation.

### CODE REVIEW

File: interfaces/seccomp/backend.go [hunk 1]
```diff
@@ -209,3 +227,56 @@ func (b *Backend) SandboxFeatures() []string {
...
+			// Technically, core16's glibc is new enough, but it
+			// always had socketcall in the template, so ensure
+			// backwards compatibility.
+			if baseSnap == "" || baseSnap == "core" || baseSnap == "core16" {
+				needed = true
+			}
+		}
+	case "powerpc", "ppc64el", "s390x":
+		// TBD
+		needed = true
```
Comment (jdstrand) [inline interfaces/seccomp/backend.go]: FYI, I've thus far been unable to test these architectures, so I checked unistd.h on these 3 architectures, and unlike i386, they have the individual socket calls defined. We could probably just default to `needed = false`. Another option is on these, add the above baseSnap check to ensure existing installs aren't affected, then see how it goes with core18.

Comment (jdstrand) [reply interfaces/seccomp/backend.go]: Ah, reading the glibc sources, we see that ./sysdeps/unix/sysv/linux/kernel-features.h says __ASSUME_SOCKETCALL needs to be defined on the architectures that support socketcall(). Looking for that, I see that:
 * sysdeps/unix/sysv/linux/powerpc/kernel-features.h: powerpc has individual socket syscalls as of 2.6.37
 * ./sysdeps/unix/sysv/linux/s390/kernel-features.h: s390 has individual socket syscalls as of 4.3
 * ./sysdeps/unix/sysv/linux/m68k/kernel-features.h: m68k has individual socket syscalls as of 4.3
 * ./sysdeps/unix/sysv/linux/sh/kernel-features.h: sh has individual socket syscalls as of 2.6.37
 * ./sysdeps/unix/sysv/linux/microblaze/kernel-features.h: microblaze has all individual socket syscalls as of 2.6.37

I'll adjust the PR accordingly.

Comment (jdstrand) [reply interfaces/seccomp/backend.go]: Done.

File: interfaces/seccomp/backend_test.go [hunk 2]
```diff
@@ -392,3 +420,103 @@ func (s *backendSuite) TestSandboxFeatures(c *C) {
...
+				}
+				// specify "core18" here so as not to influence
+				// the release check.
+				c.Assert(seccomp.RequiresSocketcall("core18"), Equals, expected)
+			}
+		}
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallForcedViaKernelVersion(c *C) {
```
Comment (mvo5) [inline interfaces/seccomp/backend_test.go]: Same question as above, e.g.:
```
func (s *backendSuite) TestRequiresSocketcallForcedViaKernelVersion(c *C) {
	restore := seccomp.MockReleaseInfoId("other")
	defer restore()

	base := "core18"
	tests := []struct {
		arch               string
		kernelVer          string
		requiresSocketcall bool
	}{
		// on i386 we only need socketcall for 4.2 (or lower) kernels
		{"i386", "4.2", true},
		{"i386", "4.3", false},
		{"i386", "4.4", false},

		// on s390x we only need socketcall for 4.2 (or lower) kernels
		{"s390x", "4.2", true},
		{"s390x", "4.3", false},
		{"s390x", "4.4", false},

		// on other arches we never need socketcall
		{"other", "4.2", false},
		{"other", "4.3", false},
		{"other", "4.4", false},
	}

	for _, t := range tests {
		restore = seccomp.MockUbuntuKernelArchitecture(func() string { return t.arch })
		defer restore()
		restore = seccomp.MockKernelVersion(func() string { return t.kernelVer })
		defer restore()

		// specify "core18" here so as not to influence the
		// kernel check.
		c.Check(seccomp.RequiresSocketcall(base), Equals, t.requiresSocketcall, Commentf("incorrect result for %v", t))
	}
}
```

File: interfaces/seccomp/backend_test.go [hunk 3]
```diff
@@ -392,3 +420,103 @@ func (s *backendSuite) TestSandboxFeatures(c *C) {
...
+	restore = seccomp.MockKernelVersion(func() string { return "4.3" })
+	defer restore()
+
+	testBases := []string{"", "core", "core16"}
+	for _, baseSnap := range testBases {
+		c.Assert(seccomp.RequiresSocketcall(baseSnap), Equals, true)
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallNotForcedViaBaseSnap(c *C) {
```
Comment (mvo5) [inline interfaces/seccomp/backend_test.go]: I was thinking about this a bit and experimenting how to make this into a single table driven tests with wildcards. Below is what I came up with. However I'm not super happy with the result, it does feel a bit too magic and non-obvious as a test (I like the table though but the implementation is a bit hard to read). So probably not worth adding.
```
func (s *backendSuite) TestRequiresSocketcall(c *C) {
	tests := []struct {
		arch    string
		distro  string
		release string
		kver    string
		base    string
	}{
		// for i386/s390x on ubuntu 14.04 we always need socketcall
		{"i386", "ubuntu", "14.04", "*", "*"},
		{"s390x", "ubuntu", "14.04", "*", "*"},
		// for i386/s390x with kernel < 4.2 we always need socketcall
		{"i386", "*", "*", "4.2", "*"},
		{"s390x", "*", "*", "4.2", "*"},
		// for i386/s390x with bases "", "core", "core16" we always need socketcall
		{"i386", "*", "*", "*", ""},
		{"s390x", "*", "*", "*", ""},
		{"i386", "*", "*", "*", "core"},
		{"s390x", "*", "*", "*", "core"},
		{"i386", "*", "*", "*", "core16"},
		{"s390x", "*", "*", "*", "core16"},
		// for sparc/sparc64 we always need socketcall
		{"sparc", "*", "*", "*", "*"},
		{"sparc64", "*", "*", "*", "*"},
	}

	// everything else is negative
	requiresSocketcall := func(arch, distro, release, kver, base string) bool {
		for _, t := range tests {
			if (t.arch == arch || t.arch == "*") && (t.distro == distro || t.distro == "*") && (t.release == release || t.release == "*") && (t.kver == kver || t.kver == "*") && (t.base == base || t.base == "*") {
				return true
			}
		}
		return false
	}

	reqs := make([]string, 0, 200)
	for _, arch := range []string{"i386", "s390x", "amd64", "armhf", "arm64", "powerpc", "ppc64el", "sparc", "sparc64", "arch-other"} {
		for _, distro := range []string{"ubuntu", "distro-other"} {
			for _, release := range []string{"14.04", "rel-other"} {
				for _, kver := range []string{"4.2", "4.3", "kver-other"} {
					for _, base := range []string{"", "core", "core16", "base-other"} {
						restore := seccomp.MockUbuntuKernelArchitecture(func() string { return arch })
						defer restore()
						restore = seccomp.MockReleaseInfoId(distro)
						defer restore()
						restore = seccomp.MockReleaseInfoVersionId(release)
						defer restore()
						restore = seccomp.MockKernelVersion(func() string { return kver })
						defer restore()

						expected := requiresSocketcall(arch, distro, release, kver, base)
						c.Check(seccomp.RequiresSocketcall(base), Equals, expected, Commentf("incorrect result for %v", []string{arch, distro, release, kver, base}))
						if expected {
							reqs = append(reqs, fmt.Sprintf("%s/%s/%s/%s/%s", arch, distro, release, kver, base))
						}
					}
				}
			}
		}
	}
	// sanity check
	c.Check(reqs, testutil.Contains, "i386/ubuntu/14.04/kver-other/base-other")
	c.Check(reqs, testutil.Contains, "s390x/ubuntu/14.04/kver-other/base-other")
	c.Check(reqs, testutil.Contains, "i386/ubuntu/rel-other/4.2/base-other")
	c.Check(reqs, testutil.Contains, "s390x/ubuntu/rel-other/4.2/base-other")
	c.Check(reqs, testutil.Contains, "sparc/distro-other/rel-other/kver-other/base-other")
	c.Check(reqs, testutil.Contains, "sparc64/distro-other/rel-other/kver-other/base-other")

	c.Check(reqs, Not(testutil.Contains), "i386/ubuntu/rel-other/other-kver/base-other")
}
```

**Comment by mvo5** [comment]
The full diff of my experiments is at http://paste.ubuntu.com/p/BQy2f9zNgg/

File: interfaces/seccomp/backend.go [hunk 4]
```diff
@@ -209,3 +227,71 @@ func (b *Backend) SandboxFeatures() []string {
...
+//   - otherwise (ie, if new enough kernel, not 14.04, and a non-16 base
+//     snap), don't force use of socketcall()
+// - if the kernel architecture is not any of the above, force the use of
+//   socketcall()
+func RequiresSocketcall(baseSnap string) bool {
+	var needed bool
+
+	switch ubuntuKernelArchitecture() {
+	case "i386", "s390x":
+		needed = false
```
[the remark this answers is not in the document]
Comment (jdstrand) [reply interfaces/seccomp/backend.go]: At one point it was needed, but then I did a cleanup and it wasn't any more. Thanks for pointing this out.

File: interfaces/seccomp/backend_test.go (hunk 3 above)
Comment (jdstrand) [reply interfaces/seccomp/backend_test.go]: Not strictly opposed to this, but the remaining tests are very small and easy to read, so I think I'd prefer not to add the magic.

Comment (mvo5) [reply interfaces/seccomp/backend_test.go]: Yeah, I agree, as I said earlier I was getting carried away a bit and there it was definitely too far :)

File: interfaces/seccomp/backend.go [hunk 5]
```diff
@@ -143,25 +150,32 @@ func (b *Backend) Remove(snapName string) error {
 // deriveContent combines security snippets collected from all the interfaces
 // affecting a given snap into a content map applicable to EnsureDirState.
 func (b *Backend) deriveContent(spec *Specification, opts interfaces.ConfinementOptions, snapInfo *snap.Info) (content map[string]*osutil.FileState, err error) {
+	addSocketcall := false
+	// Some base snaps and systems require the socketcall() in the default
+	// template
+	if requiresSocketcall(snapInfo.Base) {
```
Comment (mvo5) [inline interfaces/seccomp/backend.go]: Sorry, I think I missed that earlier. Could we simply write: `addSocketcal := requiresSocketcall(snapInfo.Base)` here?

Comment (jdstrand) [reply interfaces/seccomp/backend.go]: Yeah, that's fine. It made more sense in an earlier iteration.

**Comment by jdstrand** [comment]
Thanks!

**Pull request merged**